src/io/unix/iouring.rs

```diff
@@ -0,0 +1,176 @@
+use super::sync::{File as SyncFile, IODriver as SyncDriver};
+use crate::prelude::*;
+use bytes::{Bytes, BytesMut};
+use rio::Rio;
+use std::time::SystemTime;
+
+/// IO driver for file operations with optional async support
+#[derive(Debug, Clone)]
+pub struct IODriver {
+    rio: Option<Arc<Rio>>,
+    sync_driver: SyncDriver,
+}
+
+impl IODriver {
+    /// Create new iodriver with optional async support
+    pub fn new(rio: Option<Rio>) -> Self {
+        Self {
+            rio: rio.map(Arc::new),
+            sync_driver: SyncDriver::default(),
+        }
+    }
+
+    pub(crate) async fn open(&self, path: impl AsRef<Path>) -> IOResult<File> {
+        let sync = self.sync_driver.open(path).await?;
+        File::from_file(sync, self.rio.clone()).await
+    }
+
+    pub(crate) async fn create(&self, path: impl AsRef<Path>) -> IOResult<File> {
+        let sync = self.sync_driver.create(path).await?;
+        File::from_file(sync, self.rio.clone()).await
+    }
+}
+
+impl Default for IODriver {
+    fn default() -> Self {
+        let rio = rio::new().ok();
+        Self::new(rio)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct File {
+    sync: SyncFile,
+    rio: Option<Arc<Rio>>,
+}
+
+impl File {
+    pub(crate) fn size(&self) -> u64 {
+        self.sync.size()
+    }
+
+    pub(crate) async fn write_append_all(&self, buf: Bytes) -> IOResult<()> {
+        if let Some(ref rio) = self.rio {
+            let offset = self.sync.size.fetch_add(buf.len() as u64, Ordering::SeqCst);
+            self.write_all_at_aio(offset, buf, rio).await
+        } else {
+            self.sync.write_append_all(buf).await
+        }
+    }
+
+    pub(crate) async fn write_append_all_buffers(
+        &self,
+        first_buf: Bytes,
+        second_buf: Bytes,
+    ) -> IOResult<()> {
+        if let Some(ref rio) = self.rio {
+            let first_len = first_buf.len() as u64;
+            let second_len = second_buf.len() as u64;
+            let mut offset = self
+                .sync
+                .size
+                .fetch_add(first_len + second_len, Ordering::SeqCst);
+            self.write_all_at_aio(offset, first_buf, rio).await?;
+            offset = offset + first_len;
+            self.write_all_at_aio(offset, second_buf, rio).await
+        } else {
+            self.sync
+                .write_append_all_buffers(first_buf, second_buf)
+                .await
+        }
+    }
+
+    pub(crate) async fn write_all_at(&self, offset: u64, buf: Bytes) -> IOResult<()> {
```

Comment:
Please, add `debug_assert!(offset + buf.len() <= self.size());`. We should be sure that this function will not change file size